tokio/src/sync/broadcast.rs

```diff
@@ -661,35 +664,61 @@ impl<T> Receiver<T> {
             // the slot lock.
             drop(slot);
 
-            let tail = self.shared.tail.lock().unwrap();
+            let mut tail = self.shared.tail.lock().unwrap();
 
             // Acquire slot lock again
             slot = self.shared.buffer[idx].read().unwrap();
 
-            // `tail.pos` points to the slot that the **next** send writes to. If
-            // the channel is closed, the previous slot is the oldest value.
-            let mut adjust = 0;
-            if tail.closed {
-                adjust = 1
-            }
-            let next = tail
-                .pos
-                .wrapping_sub(self.shared.buffer.len() as u64 + adjust);
+            // Make sure the position did not change. This could happen in the
+            // unlikely event that the buffer is wrapped between dropping the
+            // read lock and acquiring the tail lock.
+            if slot.pos != self.next {
+                let next_pos = slot.pos.wrapping_add(self.shared.buffer.len() as u64);
+
+                if next_pos == self.next {
+                    // Store the waker
+                    if let Some((waiter, waker)) = waiter {
+                        // Safety: called while locked.
+                        unsafe {
+                            // Only queue if not already queued
+                            waiter.with_mut(|ptr| {
+                                if !(*ptr).queued {
+                                    (*ptr).queued = true;
+                                    (*ptr).waker = Some(waker.clone());
+                                    tail.waiters.push_front(NonNull::new_unchecked(&mut *ptr));
+                                }
+                            });
+                        }
+                    }
+
+                    return Err(TryRecvError::Empty);
+                }
 
-            let missed = next.wrapping_sub(self.next);
+                // `tail.pos` points to the slot that the **next** send writes to. If
+                // the channel is closed, the previous slot is the oldest value.
+                let mut adjust = 0;
```

Comment:
Even with the comments I don't really understand how we get here. What is the condition?
There is no new value available yet, but the sender is closed? Then we should never need to wait?

Reply:
Hmmm? Not sure I follow. You mean how do we get to the bit of code starting with `let mut adjust = 0`?

Reply:
I basically have no idea what scenario this part of the function is handling and why we need adjust/closed/etc 😀

Reply:
This is from before this change, but I will try to clarify the comment.

Reply:
I updated the comment. Does it make more sense?

Reply:
Thanks! Makes more sense now. However I'm wondering whether just going to the newest position wouldn't be easier and more helpful (if you already lagged it's likely it might happen again. And starting with the max amount of buffered elements might give you a bit more runaway time until it happens than starting at the boundary).

Reply:
Receivers can't just be dropped. They have to iterate each remaining slot to decrement the remaining count. If the user wishes to get the latest, on error, they can drop the receiver and create a new one. This would be equivalent. I'm happy to revisit the details (i'm not entirely satisfied w/ the broadcast channel) but in 0.2 we must maintain behavior.